File: android/src/main/java/xyz/luan/audioplayers/WrappedSoundPool.java
package xyz.luan.audioplayers;

import android.media.AudioAttributes;
import android.media.AudioManager;
import android.media.SoundPool;
import android.os.Build;
import android.os.PowerManager;
import android.content.Context;
import android.util.Log;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.net.URL;
import java.util.Map;
import java.util.HashMap;
import java.util.Collections;

import static java.io.File.createTempFile;

public class WrappedSoundPool extends Player {

    private static SoundPool soundPool = createSoundPool();
    static {
        soundPool.setOnLoadCompleteListener(new SoundPool.OnLoadCompleteListener() {
            public void onLoadComplete(SoundPool soundPool, int sampleId, int status) {
                Log.d("WSP", "Loaded " + sampleId);
                WrappedSoundPool player = soundIdToPlayer.get(sampleId);
                if (player != null) {
                    player.loading = false;
                    if (player.playing) {
                        player.start();
                    }
                }
            }
        });
    }

    private static Map<Integer, WrappedSoundPool> soundIdToPlayer = Collections.synchronizedMap(new HashMap<Integer, WrappedSoundPool>());


    private final AudioplayersPlugin ref;

    private final String playerId;

    private String url;

    private float volume = 1.0f;

    private Integer soundId;

    private Integer streamId;

    private boolean playing = false;

    private boolean paused = false;

    private boolean looping = false;

    private boolean loading = false;

    WrappedSoundPool(AudioplayersPlugin ref, String playerId) {
        this.ref = ref;
        this.playerId = playerId;
    }

    @Override
    String getPlayerId() {
        return playerId;
    }

    @Override
    void play() {
        if (!this.loading) {
            start();
        }
        this.playing = true;
    }

    @Override
    void stop() {
        if (this.playing) {
            soundPool.stop(this.streamId);
            this.playing = false;
        }
        this.paused = false;
    }

    @Override
    void release() {
        this.stop();
        if (this.soundId != null) {
            soundPool.unload(this.soundId);
            soundIdToPlayer.remove(this.soundId);
            this.soundId = null;
        }
    }

    @Override
    void pause() {
        if (this.playing) {
            soundPool.pause(this.streamId);
            this.playing = false;
            this.paused = true;
        }
    }

    @Override
    void setUrl(final String url, final boolean isLocal) {
        if (this.url != null && this.url.equals(url)) {
            return;
        }

        if (this.soundId != null) {
            release();
        }

        this.url = url;
        this.loading = true;

        // TODO Not sure that start a thread for each load is the sane way to go.
        new Thread(new Runnable() {
            @Override
            public void run() {
                final WrappedSoundPool self = WrappedSoundPool.this;
                self.soundId = soundPool.load(getAudioPath(url, isLocal), 1);
                soundIdToPlayer.put(self.soundId, self);
            }
        }).start();
    }

    @Override
    void setVolume(double volume) {
        this.volume = (float) volume;
        if (this.playing) {
            soundPool.setVolume(this.streamId, this.volume, this.volume);
        }
    }

    @Override
    void configAttributes(boolean respectSilence, boolean setWakeMode, Context context) {
    }

    @Override
    void setReleaseMode(ReleaseMode releaseMode) {
        this.looping = releaseMode == ReleaseMode.LOOP;
        if (this.playing) {
            soundPool.setLoop(streamId, this.looping ? -1 : 0);
        }
    }

    @Override
    int getDuration() {
        throw unsupportedOperation("getDuration");
    }

    @Override
    int getCurrentPosition() {
        throw unsupportedOperation("getCurrentPosition");
    }

    @Override
    boolean isActuallyPlaying() {
        return false;
    }

    @Override
    void seek(int position) {
        throw unsupportedOperation("seek");
    }

    private static SoundPool createSoundPool() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            AudioAttributes attrs = new AudioAttributes.Builder().setLegacyStreamType(AudioManager.USE_DEFAULT_STREAM_TYPE)
                    .setUsage(AudioAttributes.USAGE_GAME)
                    .build();
            return new SoundPool.Builder()
                    .setAudioAttributes(attrs)
                    .setMaxStreams(100)
                    .build();
        }
        return unsafeBuildLegacySoundPool();
    }

    @SuppressWarnings("deprecation")
    private static SoundPool unsafeBuildLegacySoundPool() {
        return new SoundPool(1, AudioManager.STREAM_MUSIC, 1);
    }

    private void start() {
        if (this.paused) {
            soundPool.resume(this.streamId);
            this.paused = false;
        } else {
            this.streamId = soundPool.play(
                    soundId,
                    this.volume,
                    this.volume,
                    0,
                    this.looping ? -1 : 0,
                    1.0f);
        }
    }

    private String getAudioPath(String url, boolean isLocal) {
        if (isLocal) {
            return url;
        }
        return loadTempFileFromNetwork(url).getAbsolutePath();
    }

    private File loadTempFileFromNetwork(String url) {
        FileOutputStream fileOutputStream = null;
        try {
            byte[] bytes = downloadUrl(URI.create(url).toURL());
            File tempFile = createTempFile("sound", "");
            fileOutputStream = new FileOutputStream(tempFile);
            fileOutputStream.write(bytes);
            tempFile.deleteOnExit();
            return tempFile;
        } catch (IOException e) {
            throw new RuntimeException(e);
        } finally {
            try {
                if (fileOutputStream != null) {
                    fileOutputStream.close();
                }
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
    }

    private byte[] downloadUrl(URL url) {
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();

        InputStream stream = null;
        try {
            byte[] chunk = new byte[4096];
            int bytesRead;
            stream = url.openStream();

            while ((bytesRead = stream.read(chunk)) > 0) {
                outputStream.write(chunk, 0, bytesRead);
            }

        } catch (IOException e) {
            throw new RuntimeException(e);
        } finally {
            try {
                stream.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        return outputStream.toByteArray();
    }

    private UnsupportedOperationException unsupportedOperation(String message) {
        return new UnsupportedOperationException("LOW_LATENCY mode does not support: " + message);
    }

}

File: android/src/main/kotlin/xyz/luan/audioplayers/WrappedSoundPool.kt
package xyz.luan.audioplayers

import android.content.Context
import android.media.AudioAttributes
import android.media.AudioManager
import android.media.MediaDataSource
import android.media.SoundPool
import android.util.Log
import java.io.*
import java.net.URI
import java.net.URL
import java.util.*

class WrappedSoundPool internal constructor(override val playerId: String) : Player() {
    companion object {
        private val soundPool = createSoundPool()

        /** For the onLoadComplete listener, track which sound id is associated with which player. An entry only exists until
         * it has been loaded.
         */
        private val soundIdToPlayer = Collections.synchronizedMap(HashMap<Int, WrappedSoundPool>())

        /** This is to keep track of the players which share the same sound id, referenced by url. When a player release()s, it
         * is removed from the associated player list. The last player to be removed actually unloads() the sound id and then
         * the url is removed from this map.
         */
        private val urlToPlayers = Collections.synchronizedMap(mapOf<String, MutableList<WrappedSoundPool>>())

        private fun createSoundPool(): SoundPool {
            val attrs = AudioAttributes.Builder().setLegacyStreamType(AudioManager.USE_DEFAULT_STREAM_TYPE)
                    .setUsage(AudioAttributes.USAGE_GAME)
                    .build()
            return SoundPool.Builder()
                    .setAudioAttributes(attrs)
                    .setMaxStreams(100)
                    .build()
        }

        init {
            soundPool.setOnLoadCompleteListener { _, sampleId, _ ->
                Log.d("WSP", "Loaded $sampleId")
                val loadingPlayer = soundIdToPlayer[sampleId]
                if (loadingPlayer != null) {
                    soundIdToPlayer.remove(loadingPlayer.soundId)
                    // Now mark all players using this sound as not loading and start them if necessary
                    synchronized(urlToPlayers) {
                        val urlPlayers: List<WrappedSoundPool>? = urlToPlayers[loadingPlayer.url]
                        for (player: WrappedSoundPool in urlPlayers!!) {
                            Log.d("WSP", "Marking $player as loaded")
                            player.loading = false
                            if (player.playing) {
                                Log.d("WSP", "Delayed start of $player")
                                player.start()
                            }
                        }
                    }
                }
            }
        }
    }

    private var url: String? = null
    private var volume = 1.0f
    private var rate = 1.0f
    private var soundId: Int? = null
    private var streamId: Int? = null
    private var playing = false
    private var paused = false
    private var looping = false
    private var loading = false

    override fun play(context: Context) {
        if (!loading) {
            start()
        }
        playing = true
    }

    override fun stop() {
        if (playing) {
            soundPool.stop((streamId)!!)
            playing = false
        }
        paused = false
    }

    override fun release() {
        stop()
        if (soundId != null && url != null) {
            synchronized<Unit>(urlToPlayers) {
                val playersForSoundId = urlToPlayers[url] ?: return
                if (playersForSoundId.singleOrNull() === this) {
                    urlToPlayers.remove(url)
                    soundPool.unload(soundId!!)
                    soundIdToPlayer.remove(soundId)
                    soundId = null
                    Log.d("WSP", "Unloaded soundId $soundId")
                } else {
                    // This is not the last player using the soundId, just remove it from the list.
                    playersForSoundId.remove(this)
                }
            }
        }
    }

    override fun pause() {
        if (playing) {
            soundPool.pause((streamId)!!)
            playing = false
            paused = true
        }
    }

    override fun setDataSource(mediaDataSource: MediaDataSource?, context: Context) {
        throw unsupportedOperation("setDataSource")
    }

    override fun setUrl(url: String?, isLocal: Boolean, context: Context) {
        if (this.url != null && (this.url == url)) {
            return
        }
        if (soundId != null) {
            release()
        }
        synchronized(urlToPlayers) {
            this.url = url
            var urlPlayers = urlToPlayers[url]
            if (urlPlayers != null) {
                // Sound has already been loaded - reuse the soundId.
                val originalPlayer = urlPlayers.first()
                soundId = originalPlayer.soundId
                loading = originalPlayer.loading
                urlPlayers.add(this)
                Log.d("WSP", "Reusing soundId$soundId for $url is loading=$loading $this")
                return
            }

            // First one for this URL - load it.
            loading = true
            val start = System.currentTimeMillis()
            soundId = soundPool.load(getAudioPath(url, isLocal), 1)
            Log.d("WSP", "time to call load() for $url: ${System.currentTimeMillis() - start} player=$this")
            soundIdToPlayer[soundId] = this
            urlPlayers = ArrayList()
            urlPlayers.add(this)
            urlToPlayers.put(url, urlPlayers)
        }
    }

    override fun setVolume(volume: Double) {
        this.volume = volume.toFloat()
        if (playing) {
            soundPool.setVolume((streamId)!!, this.volume, this.volume)
        }
    }

    override fun setRate(rate: Double): Int {
        this.rate = rate.toFloat()
        if (streamId != null) {
            soundPool.setRate(streamId!!, this.rate)
            return 1
        }
        return 0
    }

    override fun configAttributes(
            respectSilence: Boolean,
            stayAwake: Boolean,
            duckAudio: Boolean,
            context: Context,
    ) {}

    override fun setReleaseMode(releaseMode: ReleaseMode) {
        looping = releaseMode === ReleaseMode.LOOP
        if (playing) {
            soundPool.setLoop(streamId!!, if (looping) -1 else 0)
        }
    }

    override val duration: Int
        get() = throw unsupportedOperation("getDuration")

    override val currentPosition: Int
        get() = throw unsupportedOperation("getCurrentPosition")

    override val isActuallyPlaying: Boolean
        get() = false

    override fun setPlayingRoute(playingRoute: String, context: Context) {
        throw unsupportedOperation("setPlayingRoute")
    }

    override fun seek(position: Int) {
        throw unsupportedOperation("seek")
    }

    private fun start() {
        setRate(rate.toDouble())
        if (paused) {
            soundPool.resume((streamId)!!)
            paused = false
        } else {
            streamId = soundPool.play(
                    soundId!!,
                    volume,
                    volume,
                    0,
                    if (looping) -1 else 0,
                    1.0f,
            )
        }
    }

    private fun getAudioPath(url: String?, isLocal: Boolean): String? {
        return if (isLocal) url else loadTempFileFromNetwork(url).absolutePath
    }

    private fun loadTempFileFromNetwork(url: String?): File {
        var fileOutputStream: FileOutputStream? = null
        try {
            val bytes = downloadUrl(URI.create(url).toURL())
            val tempFile = File.createTempFile("sound", "")
            fileOutputStream = FileOutputStream(tempFile)
            fileOutputStream.write(bytes)
            tempFile.deleteOnExit()
            return tempFile
        } catch (e: IOException) {
            throw RuntimeException(e)
        } finally {
            try {
                fileOutputStream?.close()
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
    }

    private fun downloadUrl(url: URL): ByteArray {
        val outputStream = ByteArrayOutputStream()
        var stream: InputStream? = null
        try {
            val chunk = ByteArray(4096)
            var bytesRead: Int
            stream = url.openStream()
            while ((stream.read(chunk).also { bytesRead = it }) > 0) {
                outputStream.write(chunk, 0, bytesRead)
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        } finally {
            try {
                stream!!.close()
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
        return outputStream.toByteArray()
    }

    private fun unsupportedOperation(message: String): UnsupportedOperationException {
        return UnsupportedOperationException("LOW_LATENCY mode does not support: $message")
    }
}